packages/common/src/tools/LogThrottler.ts

```diff
@@ -0,0 +1,84 @@
+// eslint-disable-next-line import/no-cycle
+import { Logger } from './Logger'
+
+interface LogInfo {
+  count: number
+  throttleCount: number
+  isThrottling: boolean
+}
+
+export interface LogThrottlerOptions {
+  readonly threshold: number
+  readonly thresholdTimeInMs: number
+  readonly throttleTimeInMs: number
+}
+/**
+ * @summary Aim of this class is to reduce log output by throttling.
+ *
+ * @description It allows to throttle given log for throttleTimeInMs if it was called threshold times in throttleTimeInMs
+ * by storing count of how much time it was called in recentLogs by logKey.
+ */
+export class LogThrottler {
+  private readonly recentLogs: Map<string, LogInfo>
+
+  constructor(
+    private readonly options: LogThrottlerOptions,
+    private readonly logger: Logger,
+  ) {
+    this.recentLogs = new Map<string, LogInfo>()
+    this.logger.for(LogThrottler.name)
+    setInterval(() => this.clearCount(), options.thresholdTimeInMs)
+  }
+
+  add(logKey: string): void {
+    let logInfo = this.recentLogs.get(logKey)
+
+    if (!logInfo) {
+      logInfo = {
+        count: 0,
+        isThrottling: false,
+        throttleCount: 0,
+      }
+      this.recentLogs.set(logKey, logInfo)
+    }
+
+    this.incrementCount(logInfo)
+
+    if (!logInfo.isThrottling && logInfo.count >= this.options.threshold) {
+      this.throttle(logKey, logInfo)
+    }
+  }
+
+  isThrottling(logKey: string): boolean | undefined {
```

Comment:
I think it's much cleaner to have simply `boolean` return type here (and do !! inside the function) so callers don't have to worry about handling `undefined`.

Reply:
that's totally right